lsf/man/man8/lim.8
.\" $Id: lim.8,v 1.2 2012/08/13 21:28:01 cchen Exp $
.ds ]W %
.ds ]L
.TH LIM 8 "OpenLava Version 4.0 - Aug 2016"
.SH NAME
lim \- Load Information Manager (\s-1LIM\s0) for the OpenLava system
.SH SYNOPSIS
\fBLSF_SERVDIR/lim [ -h ] [ -V ] [ -t ] [ -C ] [ -d \fIenv_dir\fB ] [ -\fIdebug_level\fB ]
.SH DESCRIPTION
\s-1LIM\s0 is a server that runs on every server host which is
participating in load sharing. It provides system configuration
information, load information, and placement advice services.  The
\s-1LIM\s0s on all hosts coordinate in collecting and transmitting load
information. Load information is transmitted between \s-1LIM\s0s in the form of
vectors of load indices. These are described in the 
\fBLOAD INDICES\fR section.
.PP
In order to run \s-1LIM\s0 on a host, the host must be configured as a
server host in the \fBlsf.cluster.\fR<\fIclustername\fR> file. See
the description of the \fBSERVER\fR keyword in the 
\fBlsf.cluster.\fR<\fIclustername\fR> file in
.BR lsf.cluster (5)
for more information on server hosts and client hosts.
.PP
A master \s-1LIM\s0 is elected for each OpenLava cluster. The master
\s-1LIM\s0 receives load information from all slave \fBLIM\fRs
and provides services to all hosts. 
The slave \fBLIM\fRs periodically check their own load conditions and
send a load vector to the master if significant changes in load condition are
observed. The minimum load information exchange interval is 15 seconds.
However, a user site can redefine this by setting the \fBEXINTERVAL\fR parameter
in \fBPARAMETERS\fR section of \fBlsf.cluster\fR file (see
.BR lsf.cluster (5)).
The actual exchange intervals
can be bigger if the load does not change much.
.PP
Slave \s-1LIM\s0s also provide some services to the local application. For
example, applications contact the local \s-1LIM\s0 to get local load
information, the local cluster name, the available resources on any
host, and the name of the host on which the master \s-1LIM\s0 is running.
A slave \s-1LIM\s0 accepts
application requests to lock or unlock the local host.
Slave \s-1LIM\s0s also monitor the status of the master \s-1LIM\s0 and
elect a new master if the original one becomes unavailable.
.PP
\s-1LIM\s0 reads
.BR lsf.conf (5)
file to get the following parameters: \fBLSF_LIM_DEBUG\fR,
\fBLSF_CONFDIR\fR, \fBLSF_SERVERDIR\fR, \fBLSF_LIM_PORT\fR, \fBLSF_LOG_MASK\fR and \fBLSF_LOGDIR\fR.
If \fBLSF_LIM_DEBUG\fR is defined, \s-1LIM\s0 runs at the specified debug
level. \fBLSF_CONFDIR\fR tells \s-1LIM\s0 where to find the OpenLava configuration
files. \fBLSF_SERVERDIR\fR is used by \s-1LIM\s0 during reconfiguration in
order to locate the directory where the \s-1LIM\s0 binary is stored. 
\fBLSF_LIM_PORT\fR is used to define the UDP port number that \s-1LIM\s0 uses for
serving all applications. If this is not defined, \s-1LIM\s0 will try to
get the port number from services database (see 
.BR getservbyname (3)). 
If \fBLSF_LOG_MASK\fR is defined, the defined log mask is set so that messages
with lower priority than \fBLSF_LOG_MASK\fR are not logged. If \fBLSF_LOG_MASK\fR is
not defined, a default log mask \fBLOG_WARNING\fR is used. If \fBLSF_LOGDIR\fR is
defined, error messages are logged into the file
\fBlim.log.\fR<\fIhostname\fR>\fR in the directory \fBLSF_LOGDIR\fR. If \s-1LIM\s0
cannot write in \fBLSF_LOGDIR\fR, errors are logged into \fB/tmp\fR. If \fBLSF_LOGDIR\fR
is not defined, \fBsyslog\fR is used to log error messages with the level \fBLOG_ERR\fR.
\fBLSF_CLUSTER_NAME\fR is a mandatory parameter which specifies the cluster that
the \s-1LIM\s0 belongs to.
.SH LOAD INDICES
The load information provided by the \s-1LIM\s0 consists of the
following load indices:
.TP 5
.B r15s
The 15-second exponentially averaged \s-1CPU\s0 run queue length, normalized by
the \s-1CPU\s0 speed.
.TP 5
.B r1m
The 1-minute exponentially averaged \s-1CPU\s0 run queue length, normalized by
the \s-1CPU\s0 speed.
.TP 5
.B r15m
The 15-minute exponentially averaged \s-1CPU\s0 run queue length, normalized by
the \s-1CPU\s0 speed.
.TP 5
.B ut
The \s-1CPU\s0 utilization exponentially averaged over the last minute,
between 0 and 1.
.TP 5
.B pg
The memory paging rate exponentially averaged over the last minute,
in pages per second.
.TP 5
.B io
The disk I/O rate exponentially averaged over the last minute,
in KBytes per second.
.TP 5
.B ls
The number of current login users.
.TP 5
.B it
The idle time of the host (keyboard not touched on all logged in sessions),
in minutes.
.TP 5
.B tmp
The amount of free disk space in \fB/tmp\fR, in MBytes.
.TP 5
.B swp
The amount of currently available swap space, in MBytes.
.TP 5
.B mem
The amount of currently available memory, in MBytes.
.PP
In addition, an OpenLava installation can configure arbitrary
external load indices as explained below.
.SH EXTERNAL LOAD INFORMATION MANAGER
The load indices monitored at a site can be extended by directing the
\s-1LIM\s0 to invoke and communicate with an External Load Information
Manager (\s-1ELIM\s0). The \s-1ELIM\s0 is responsible for collecting load
indices not managed by the \s-1LIM\s0. These indices are passed on to the
\s-1LIM\s0 by \s-1ELIM\s0 through a well defined protocol.
.PP
To configure an external load index, the index name must first be defined
in the \fBResources\fR section of the
.BR lsf.shared (5)
file.  The location of the resource must then be defined in the
\fBResourceMap\fR section of the
.BR lsf.cluster (5)
file.
A load index is either a shared or a non-shared resource.  A non-shared
resource is defined on each host in the cluster, and the value is specific
to each host.  A shared resource is a resource whose value is shared by
more than one host, and the resource may be defined only on a subset of
the hosts.  The
\s-1ELIM\s0 can report both shared and non-shared resources.  The
location specification of the resource in the
\fBResourceMap\fR section of the
.BR lsf.cluster (5)
file defines whether a resource is shared or non-shared.
.PP
If a non-shared external index is defined, the \s-1LIM\s0 on each host
will invoke the \s-1ELIM\s0 on start up.  If a shared external index is
defined, the \s-1LIM\s0 will start the \s-1ELIM\s0 only if the index applies
to the current host, and the current host is the first host from 
the hostlist defined for the resource instance, which is alive.  
(see
.BR lsf.cluster (5)).
The executable for the \s-1ELIM\s0 must be in \fBLSF_SERVDIR\fR
and must have the name '\fBelim\fR'. 
The \s-1ELIM\s0 will run with the same user ID
and file access permissions as the \s-1LIM\s0.
If the \s-1ELIM\s0 dies, it will be
restarted by the \s-1LIM\s0. If the \s-1LIM\s0 dies, the master LIM
will ensure that the \s-1ELIM\s0 on the next host on the hostlist defined for 
the resource instance is started.  If this dead \s-1LIM\s0 comes back alive 
again however, the original \s-1ELIM\s0 will be restarted and the backup 
\s-1ELIM\s0 terminated.  
.PP
When the \s-1LIM\s0 terminates, it will send a SIGTERM signal to the \s-1ELIM\s0.
The \s-1ELIM\s0 is expected to terminate upon receiving this signal.
.PP
The \s-1LIM\s0 communicates with the \s-1ELIM\s0 through 2 environment 
variables:  \s-1LSF_MASTER\s0 and  \s-1LSF_RESOURCES\s0.  \s-1LSF_MASTER\s0
is set to an empty string if the \s-1ELIM\s0 is being started by the 
\s-1LIM\s0 on the master host.  It is left unset otherwise.
\s-1LSF_RESOURCES\s0 is set to a space separated string of dynamic 
shared resources indices for which the \s-1ELIM\s0 on that host is responsible 
to collect.  Thus if a host is defined in the resource instances 
for the dynamic shared resource indices, `\fBwork\fR', `\fBnetio\fR', and
`\fBusers\fR', then \s-1LSF_RESOURCES\s0 will be set with the string
"\fBwork\fR \fBnetio\fR \fBusers\fR".
.PP
The \s-1ELIM\s0 communicates with the \s-1LIM\s0 by periodically writing a
load update string to its stdout. The load update string
contains the number of indices followed by a list of name-value pairs
in the following format:
"\fInumIndx indexname1 value1 indexname2 value2 .... indexnameN valueN\fR".
For example "\fB3 work 47.5 netio 344.0 users 5\fR"
is a valid load update string
in which the \s-1ELIM\s0 is reporting 3 indices, `\fBwork\fR', `\fBnetio\fR', and `\fBusers\fR'
with values \fB47.5\fR, \fB344.0\fR, and \fB5\fR, respectively.
Index names corresponding to the built-in indices are also accepted.
In this case the value produced by the \s-1ELIM\s0 overrides
the value produced by \s-1LIM\s0. It is up to the \s-1ELIM\s0 to ensure
that the semantics of all indices it samples correspond to those
returned by
.BR lsinfo (1)
or
.BR ls_info (3).
.PP
The \s-1ELIM\s0 should ensure that the entire load update string
is written successfully to stdout. This can be done by checking the
return value of \fBprintf\fR(3) or \fBfprintf\fR(3) if the \s-1ELIM\s0 is implemented
as C program or the return code of 
.BR echo (1)
from a shell script.
Failure to write the load update string should cause the \s-1ELIM\s0 to
terminate.
.SH CUSTOMIZATION OF PARAMETERS
You can customize \s-1LIM\s0 by changing the configuration files in the
\fBLSF_CONFDIR\fR directory (defined in
.BR lsf.conf (5)).
The configuration file
.BR lsf.cluster (5)
located in that directory define OpenLava clusters, the resources on
individual hosts, the \s-1CPU\s0 speeds of individual hosts, whether a
host is a server host or client host, the load threshold values beyond
which a host is considered to be overloaded, the run windows during
which a host is available for load sharing, and so on.
.SH OPTIONS
.TP 5
.B -h
Print command usage to stderr and exit.
.TP 5
.B -V
Print OpenLava release version to stderr and exit.
.TP 5
.B -t
Displays host information like host architecture, CPU model, number of 
physical processors, number of cores etc. The data is from /proc/cpuinfo.
.TP 5
.B -C
Check the configuration file content with verbose error reporting
and exit.
.TP 5
.B -d \fIenv_dir\fR
Read \fBlsf.conf\fR from the directory
.I env_dir,
rather than the default directory \fB/etc\fR, or the directory specified by
the \fBLSF_ENVDIR\fR environment variable.
.TP 5
.BI - debug_level
Set the debug level. Valid values are 1 and 2. If specified, \s-1LIM\s0
runs in debugging mode. In debugging mode, \s-1LIM\s0 uses a hardcoded
port number rather than the one registered in system services. Also,
privileged operations such as reconfiguration and host lock or unlock can
be done by any user. If debug mode is not enabled, only root and
\fBLSF_MANAGER\fR (defined in
.BR lsf.conf (5))
can do these privileged operations. If \fIdebug_level\fR is 1, \s-1LIM\s0
runs in the background, with no associated control terminal. If \fIdebug_level\fR
is 2, \s-1LIM\s0 runs in the foreground, printing error messages on to
tty. The \fIdebug_level\fR option overrides the environment variable
\fBLSF_LIM_DEBUG\fR defined in
.BR lsf.conf (5).
.SH NOTE
\s-1LIM\s0 needs read access to \fB/dev/kmem\fR or its equivalent.
.SH FILES
.PD 0
.TP
\fB/etc/lsf.conf\fR (by default) or \fBLSF_ENVDIR/lsf.conf\fR
.TP
.B LSF_CONFDIR/lsf.shared
.TP
.B LSF_CONFDIR/lsf.cluster.\fR<\fIclustername\fR>
.PD
.SH "SEE ALSO"
.BR lsf.conf (5),
.BR lsf.cluster (5),
.BR lsinfo (1),
.BR ls_info (3),
.BR syslog (3)


lsf/lim/lim.load.c
/*
 * Copyright (C) 2011-2013 David Bigagli
 *
 * $Id: lim.load.c 397 2007-11-26 19:04:00Z mblack $
 * Copyright (C) 2007 Platform Computing Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
 *
 */

#include "lim.h"

#define NL_SETN 24

enum loadstruct {e_vec, e_mat};

/* Globals that can be set in the parameters section
 * in the cluster file.
 */
float  exchIntvl = EXCHINTVL;       /* 15 */
float  sampleIntvl = SAMPLINTVL;    /* 5 */
short  hostInactivityLimit = HOSTINACTIVITYLIMIT; /* 5 */
short  masterInactivityLimit = MASTERINACTIVITYLIMIT;  /* 2 */
short  resInactivityLimit    = RESINACTIVITYLIMIT;     /* 9 */
short  retryLimit = RETRYLIMIT;
short  keepTime = KEEPTIME;

time_t lastSbdActiveTime = 0;

char   mustSendLoad = TRUE;

extern int maxnLbHost;

static void rcvLoadVector (XDR *, struct sockaddr_in *, struct LSFHeader *);
static void copyResValues (struct loadVectorStruct, struct hostNode *);

void
sendLoad(void)
{
    static int noSendCount = 0;
    struct loadVectorStruct myLoadVector;
    enum   loadstruct loadType;
    struct hostNode *hPtr;
    struct sockaddr_in toAddr;
    int    i;
    int    bufSize;
    enum   limReqCode limReqCode;
    XDR    xdrs;
    char   *repBuf;
    int    sendInfo = SEND_NO_INFO;
    struct LSFHeader reqHdr;

    limReqCode = LIM_LOAD_UPD;
    resInactivityCount++;

    if (resInactivityCount > resInactivityLimit)
        myHostPtr->status[0] |= LIM_RESDOWN;

    if (time(0) - lastSbdActiveTime > SBD_ACTIVE_TIME)
        myHostPtr->status[0] |= LIM_SBDDOWN;

    if (logclass & LC_TRACE)
       ls_syslog(LOG_DEBUG, "%s: Entering ..", __func__);

    if (masterMe) {

        for (hPtr = myClusterPtr->hostList; hPtr; hPtr = hPtr->nextPtr) {

            if (hPtr == myHostPtr)
                continue;

                hPtr->hostInactivityCount++;
                if (hPtr->hostInactivityCount > 10000)
                    hPtr->hostInactivityCount = 100;

                if (! LS_ISUNAVAIL(hPtr->status)) {
                    if (hPtr->hostInactivityCount > (hostInactivityLimit + retryLimit)) {

			if (logclass & LC_COMM) {
			    ls_syslog(LOG_INFO, "\
%s: Declaring %s unavailable inactivity Count=%d", __func__,
                                  hPtr->hostName, hPtr->hostInactivityCount);
			}

                        hPtr->status[0] |= LIM_UNAVAIL;
                        hPtr->infoValid = FALSE;
                        if (hPtr->numInstances > 0) {
                            int resNo;
                            for (i = 0; i < hPtr->numInstances; i++) {
                                if (hPtr->instances[i]->updHost == NULL
                                      || hPtr->instances[i]->updHost != hPtr)
                                    continue;
                                resNo = resNameDefined(hPtr->instances[i]->resName);
                                if (allInfo.resTable[resNo].flags & RESF_DYNAMIC) {
                                    strcpy (hPtr->instances[i]->value, "-");
                                    hPtr->instances[i]->updHost = NULL;
                                }
                            }
                        }
                        hPtr->loadMask  = 0;
                        hPtr->infoMask  = 0;
                    }
                    if ( (hPtr->hostInactivityCount > hostInactivityLimit) &&
                         (hPtr->hostInactivityCount <= (hostInactivityLimit + retryLimit))) {
                        if (logclass & LC_COMM) {
                            ls_syslog(LOG_DEBUG3,
                              "%s: Asking %s to send load info %d %d", __func__,
                              hPtr->hostName, hPtr->hostInactivityCount,
                              hostInactivityLimit + retryLimit);
                        }
                        announceMasterToHost(hPtr, SEND_LOAD_INFO);
                    }
                }
        }

    } else {

        myClusterPtr->masterInactivityCount++;

	if (logclass & LC_COMM) {
	    ls_syslog (LOG_INFO, "\
%s: masterInactivityCount=%d, hostInactivityLimit=%d, masterKnown=%d, retryLimit=%d",
		       __func__, myClusterPtr->masterInactivityCount,
		       hostInactivityLimit,  myClusterPtr->masterKnown,
		       retryLimit);
	}

        if (myClusterPtr->masterInactivityCount > hostInactivityLimit) {

            if (myClusterPtr->masterKnown
                && myClusterPtr->masterInactivityCount > hostInactivityLimit
                && (myClusterPtr->masterInactivityCount <= hostInactivityLimit
		    + retryLimit)) {
                ls_syslog(LOG_WARNING, "\
%s: Attempting to probe master %s", __func__,
                          myClusterPtr->masterPtr->hostName);

                mustSendLoad = TRUE;
                sendInfo = SEND_MASTER_ANN;
            }

            if (myClusterPtr->masterKnown
                && (myClusterPtr->masterInactivityCount
                    > hostInactivityLimit + retryLimit)) {

                myClusterPtr->masterPtr->status[0] = LIM_UNAVAIL;
                myClusterPtr->masterKnown  = FALSE;
                myClusterPtr->prevMasterPtr = myClusterPtr->masterPtr;
                myClusterPtr->masterPtr = NULL;
                ls_syslog(LOG_INFO, "%s: Master LIM unknown now", __func__);
                if (limParams[LIM_COMPUTE_ONLY].paramValue)
                    ls_syslog(LOG_INFO, "\
%s: compute only host will not try to take over mastership", __func__);
            }

            /* Try to take over the current master only if
             * I am not a compute only server and the master inactivity
             * counter has reached its limit times my position in
             * the cluster file.
             */
            if (limParams[LIM_COMPUTE_ONLY].paramValue == NULL
                && (myClusterPtr->masterInactivityCount > hostInactivityLimit
                    + myHostPtr->hostNo * masterInactivityLimit)) {

                if (probeMasterTcp(myClusterPtr) < 0) {
                    /* Let's do a TCP connect towards the
                     * last known master. This is kinda half
                     * baked because if the master is hanging
                     * the TCP connect is going to succeed
                     * fooling me into believing the master
                     * is all right, the cluster is going
                     * to stall.
                     */
                    initNewMaster();
                    return;
                }
                myClusterPtr->masterInactivityCount = 0;
                myClusterPtr->masterKnown  = TRUE;
                myClusterPtr->masterPtr = myClusterPtr->prevMasterPtr;
            }
        }

        if (!myClusterPtr->masterKnown)
            return;
    }

    if (!mustSendLoad) {
        for (i = 0; i < allInfo.numIndx; i++) {
            if (fabs(myHostPtr->loadIndex[i] - li[i].valuesent)
                > li[i].exchthreshold) {
                mustSendLoad = TRUE;
                break;
            }
        }
    }

    if (!mustSendLoad && noSendCount < hostInactivityLimit - 2) {
        noSendCount++;
        return;
    }

    for (i = 0; i < allInfo.numIndx; i++)
        li[i].valuesent = myHostPtr->loadIndex[i];

    if (!masterMe) {

        loadType = e_vec;
        myLoadVector.hostNo = myHostPtr->hostNo;
        myLoadVector.status = myHostPtr->status;
        myLoadVector.seqNo  = loadVecSeqNo++;
        myLoadVector.checkSum = myClusterPtr->checkSum;
        myLoadVector.flags = sendInfo;
        myLoadVector.numIndx   = allInfo.numIndx;
        myLoadVector.numUsrIndx = allInfo.numUsrIndx;
        myLoadVector.numResPairs = myHostPtr->numInstances;

        if (myLoadVector.numResPairs > 0) {
            if ((myLoadVector.resPairs  = getResPairs (myHostPtr)) == NULL) {
                ls_syslog(LOG_ERR, I18N_FUNC_FAIL, __func__, "getResPairs");
                return;
            }
        } else
            myLoadVector.resPairs = NULL;
        myLoadVector.li = myHostPtr->loadIndex;
        bufSize = sizeof (struct loadVectorStruct)
                  + allInfo.numIndx *sizeof (float)
                  + GET_INTNUM(allInfo.numIndx) * sizeof (int)
                  + myLoadVector.numResPairs * sizeof (struct resPair)
                  + 100;
        for (i = 0; i < myLoadVector.numResPairs; i++ ){
            bufSize += ALIGNWORD_(strlen(myLoadVector.resPairs[i].name) * sizeof(char) + 1) + 4;
            bufSize += ALIGNWORD_(strlen(myLoadVector.resPairs[i].value) * sizeof(char) + 1) + 4;
        }

        if (bufSize > MSGSIZE) {
            ls_syslog(LOG_ERR, "\
%s: message bigger then receive buf(%d)", __func__, bufSize);
            return;
        }

        if ((repBuf = malloc(bufSize)) == NULL) {
            ls_syslog(LOG_ERR, I18N_FUNC_FAIL_M, __func__, "malloc");
            return;
        }

        xdrmem_create(&xdrs, repBuf, bufSize, XDR_ENCODE);
        initLSFHeader_(&reqHdr);
        reqHdr.opCode  = (short) limReqCode;
        reqHdr.refCode =  0;

        if (!(xdr_LSFHeader(&xdrs, &reqHdr)
              && xdr_enum(&xdrs, (int *) &loadType)
              && xdr_loadvector(&xdrs, &myLoadVector, &reqHdr))) {
            ls_syslog(LOG_ERR, I18N_FUNC_FAIL, __func__, "xdr_enum/xdr_loadvector");
            xdr_destroy(&xdrs);
            FREEUP (repBuf);
            return;
        }

        toAddr.sin_family = AF_INET;
        toAddr.sin_port   = lim_port;

        memcpy(&toAddr.sin_addr.s_addr,
               &myClusterPtr->masterPtr->addr[0],
               sizeof(in_addr_t));

        if (logclass & LC_COMM)
            ls_syslog(LOG_INFO, "\
sendLoad: sending to %s (len=%d,port=%d)",
                      sockAdd2Str_(&toAddr), XDR_GETPOS(&xdrs),
                      lim_port);

        if (chanSendDgram_(limSock, repBuf, XDR_GETPOS(&xdrs), &toAddr) < 0) {
            ls_syslog(LOG_ERR, I18N_FUNC_S_FAIL_M, __func__, "chanSendDgram_",
                      sockAdd2Str_(&toAddr));
            xdr_destroy(&xdrs);
            FREEUP (repBuf);
            return;
        }
        xdr_destroy(&xdrs);
        FREEUP (repBuf);
    }

    mustSendLoad = FALSE;
    noSendCount = 0;
}

struct resPair *
getResPairs(struct hostNode *hPtr)
{
    int i;
    static struct resPair *resPairs;

    FREEUP(resPairs);

    if (hPtr->numInstances > 0) {
        resPairs = calloc(hPtr->numInstances,
                          sizeof(struct resPair));
        if (resPairs == NULL) {
            ls_syslog(LOG_ERR, "\
%s: malloc %dbytes failed %m", __func__,
                      hPtr->numInstances * sizeof(struct resPair));
            return NULL;
        }
    }

    for (i = 0; i <  hPtr->numInstances; i++) {
        resPairs[i].name = hPtr->instances[i]->resName;
        resPairs[i].value = hPtr->instances[i]->value;
    }

    return resPairs;
}

void
rcvLoad(XDR *xdrs, struct sockaddr_in *from, struct LSFHeader *hdr)
{
    enum   loadstruct loadType;

    if (0 && from->sin_port != lim_port) {
        ls_syslog(LOG_ERR, "\
%s: Update not from LIM: %s, expected %d",
                  __func__, sockAdd2Str_(from), ntohs(lim_port));
        return;
    }

    if (!xdr_enum(xdrs, (int *) &loadType)) {
        ls_syslog(LOG_ERR, "%s: xdr_enum failed", __func__);
        return;
    }

    if (loadType != e_vec) {
        ls_syslog(LOG_ERR, "\
%s: Invalid load type %d from host %s",
                  __func__, loadType, sockAdd2Str_(from));
        return;
    }

    rcvLoadVector(xdrs, from, hdr);
}


static void
rcvLoadVector(XDR *xdrs, struct sockaddr_in *from, struct LSFHeader *hdr)
{
    static int checkSumMismatch;
    static struct loadVectorStruct *loadVector;
    struct hostNode *hPtr;
    int i;
    int masterLock = FALSE;

    if (loadVector == NULL) {
        loadVector = calloc(1, sizeof(struct loadVectorStruct));
        loadVector->li = calloc(allInfo.numIndx, sizeof(float));
        loadVector->status = calloc((1 + GET_INTNUM(allInfo.numIndx)),
                                    sizeof(int));
    }

    if (!xdr_loadvector(xdrs, loadVector, hdr)) {
        ls_syslog(LOG_ERR, "\
%s: Error in xdr_loadvector from %s", __func__, sockAdd2Str_(from));
        return;
    }

    if (!masterMe) {
        ls_syslog(LOG_DEBUG, "\
%s: %s thinks I am the master, but I'm not",
                  __func__, sockAdd2Str_(from));
        return;
    }

    if (myClusterPtr->checkSum != loadVector->checkSum
        && checkSumMismatch < 5
        && (limParams[LSF_LIM_IGNORE_CHECKSUM].paramValue == NULL)) {
        ls_syslog(LOG_DEBUG, "\
%s: Sender (%s) may have different config?.",
                  __func__, sockAdd2Str_(from));
        checkSumMismatch++;
    }

    hPtr = findHostbyAddr(from, (char *)__func__);
    if (hPtr == NULL) {
        ls_syslog(LOG_ERR, "\
%s: Received load update from unknown host %s",
                  __func__, sockAdd2Str_(from));
        return ;
    }

    if (findHostbyList(myClusterPtr->hostList, hPtr->hostName) == NULL) {
        ls_syslog(LOG_ERR, "\
%s: Got load from client-only host %s.  Kill LIM on %s",
                  __func__, sockAdd2Str_(from), sockAdd2Str_(from));
        return;
    }

    if (hPtr->infoValid != TRUE) {
        return;
    }

    if (logclass & LC_COMM) {
	ls_syslog(LOG_INFO,"\
%s: Received load update from host %s", __func__, hPtr->hostName);
    }

    hPtr->hostInactivityCount = 0;

    if (hPtr->status[0] & LIM_LOCKEDM) {
        masterLock = TRUE;
    }

    hPtr->status[0] = loadVector->status[0];
    if (masterLock) {
        hPtr->status[0] |= LIM_LOCKEDM;
    } else {
        hPtr->status[0] &= ~LIM_LOCKEDM;
    }

    for (i = 0; i < GET_INTNUM(allInfo.numIndx); i++)
        hPtr->status[i + 1] = loadVector->status[i + 1];

    hPtr->loadMask  = 0;

    if (loadVector->seqNo - hPtr->lastSeqNo > 2
        && loadVector->seqNo > hPtr->lastSeqNo
        && hPtr->lastSeqNo != 0)

        ls_syslog(LOG_ERR, "\
%s: host %s lastSeqNo=%d seqNo=%d. Packets being dropped?",
                  __func__, hPtr->hostName,
                  hPtr->lastSeqNo, loadVector->seqNo);
    hPtr->lastSeqNo = loadVector->seqNo;

    copyResValues (*loadVector, hPtr);
    copyIndices(loadVector->li,
                loadVector->numIndx,
                loadVector->numUsrIndx,
                hPtr);

    if (loadVector->flags & SEND_MASTER_ANN)  {
        ls_syslog(LOG_INFO, "\
%s: Sending master announce to %s", __func__, hPtr->hostName);
        announceMasterToHost(hPtr, SEND_NO_INFO);
    }
}

static void
copyResValues(struct loadVectorStruct loadVector, struct hostNode *hPtr)
{
    static char fname[] = "copyResValues";
    int i, j, updHostNo, curHostNo;
    struct sharedResource *resource;
    struct resourceInstance *instance;
    struct hostNode *hostPtr;
    char *temp;

    if (loadVector.numResPairs <= 0)
        return;

    for (i = 0; i < loadVector.numResPairs; i++) {
        if ((resource = inHostResourcs(loadVector.resPairs[i].name)) == NULL) {
            ls_syslog(LOG_DEBUG2, "%s: Resource name <%s> reported by host <%s> is not in shared resource list", fname, loadVector.resPairs[i].name, hPtr->hostName);
            continue;
        }
        if ((instance = isInHostList(resource, hPtr->hostName)) == NULL) {
            ls_syslog(LOG_DEBUG2, "%s: Host <%s> does not have the resource <%s> defined", fname, hPtr->hostName, loadVector.resPairs[i].name);
            continue;
        }
        if (!strcmp(loadVector.resPairs[i].value, "-")
             && !strcmp(instance->value, "-"))
            continue;
        if (instance->updHost == NULL)
            hostPtr = hPtr;
        else {
            updHostNo = -1;
            curHostNo = -1;
            for (j = 0; j < instance->nHosts; j++) {
                if (instance->updHost == instance->hosts[j])
                    updHostNo = j;
                if (instance->hosts[j] == hPtr)
                    curHostNo = j;
                if (curHostNo >= 0 && updHostNo >= 0)
                    break;
            }
            if (curHostNo < 0 || updHostNo < 0)
                continue;
            hostPtr = findHostbyList(myClusterPtr->hostList,
                                instance->hosts[curHostNo]->hostName);
            if (hostPtr == NULL)
                continue;
            if (updHostNo < curHostNo
                     && (!strcmp (loadVector.resPairs[i].value, "-")
                     || strcmp (instance->value, "-")))

                continue;

            if (updHostNo > curHostNo
                   && !strcmp (loadVector.resPairs[i].value, "-"))

                continue;
            if (updHostNo == curHostNo && instance->value
                && loadVector.resPairs[i].value
                && !strcmp (loadVector.resPairs[i].value, instance->value))
                continue;
        }
        if ((temp = putstr_(loadVector.resPairs[i].value)) == NULL) {
            ls_syslog(LOG_ERR, I18N_FUNC_FAIL_M, fname, "malloc");
             return;
        }
        FREEUP (instance->value);
        instance->value = temp;
        instance->updHost = hostPtr;
    }
}

void
copyIndices(float *lindx, int numIndx, int numUsrIndx, struct hostNode *hPtr)
{
    int myBuiltIn, slaveBuiltIn, i;

    myBuiltIn = allInfo.numIndx - allInfo.numUsrIndx;
    slaveBuiltIn = numIndx - numUsrIndx;


    for (i = 0; (i < myBuiltIn) && (i < slaveBuiltIn); i++) {
        int nprocs = hPtr->statInfo.maxCpus;
        float cpuf = (hPtr->hModelNo >= 0) ?
            shortInfo.cpuFactors[hPtr->hModelNo] : 1.0;
        float rawql;

        if (i==R15S || i==R1M || i==R15M ) {
            rawql = lindx[i];
            hPtr->loadIndex[i] = normalizeRq(rawql,cpuf,nprocs);
            hPtr->uloadIndex[i] = rawql;
        } else {
            hPtr->loadIndex[i] = lindx[i];
            hPtr->uloadIndex[i] = lindx[i];
        }
    }


     for(; i < myBuiltIn; i++) {
         hPtr->loadIndex[i]  = INFINIT_LOAD;
         hPtr->uloadIndex[i] = INFINIT_LOAD;
     }


     for(i=0; (i < numUsrIndx) &&
              (i < allInfo.numUsrIndx); i++) {
         hPtr->loadIndex[myBuiltIn + i] = lindx[slaveBuiltIn + i];
         hPtr->uloadIndex[myBuiltIn + i] = lindx[slaveBuiltIn + i];
     }
}

float
normalizeRq(float rawql, float cpuFactor, int nprocs)
{
    float nrq;
    float f,k1,k2;
    float slope;
    int ifloor;

    if (rawql < 0)
        return(0.0);

    if (rawql >= INFINIT_LOAD)
        return(rawql);

    if (nprocs >= 1) {
        ifloor = rawql/nprocs;
        k1 = (rawql - nprocs*ifloor)/nprocs;
        k2 = MIN(k1,0.9);
        slope = 1.0-1.0/nprocs;
        k2 = MIN(k2,slope);
        slope = MIN(slope,0.9);
        f  = slope*(-k1 + 1.0)/(1.0 - k2);
    } else {
        return((rawql + 1) / cpuFactor);
    }

    nrq = (f * ifloor + (1.0-f) * rawql/nprocs + 1) / cpuFactor;
    if (nrq < 0)
        return (0.0);

    return (nrq);
}
